plutus-scb/src/Plutus/SCB/Events/Contract.hs

```diff
@@ -1,92 +1,216 @@
 -- The events that compiled Plutus contracts are concerned with. For each type
 -- of event there is a request constructor in 'ContractRequest' and a response
 -- constructor in 'ContractResponse'.
+
 {- Note [Contract Events]
 
 Each contract instance has two event queues in the SCB: One for the requests
-it makes, typed 'ContractRequestEvent', and one for the responses it receives,
-typed 'ContractResponseEvent'.
+it makes, typed 'ContractOutboxMessage', and one for the responses it receives,
+typed 'ContractInboxMessage'.
+
+-}
+
+{- Note [Contract Iterations]
+
+Contracts are executed iteratively:
 
-The contract instances themselves deal with *sets of active endpoints* instead of
-requests and responses directly. When the set of active endpoints of an instance
-changes, the SCB needs to translate the change into 'ContractRequestEvent' values
-by creating 'IssueRequest' events for the endpoints that have been added (compared
-to the previous set of active endpoints) and creating 'CancelRequest' values for
-endpoints that have disappeared.
+* We look at the current state of the contract to see which requests (hooks)
+  it wants to make
+* We process one of those requests and feed the result to the contract
+* The contract produces a new state, which we again inspect to get the hooks
+  that are currently active
+
+In each iteration we only feed one event to the contract, so if there is more
+than one hook then the other ones will be un-answered.  If the contract had
+been waiting for the first of several possible events, then the other hooks
+will disappear from the set of active hooks. But if it had been waiting for
+all of those events to happen then the remaining hooks will appear in the
+result, and we handle them in the next iteration.
+
+So each iteration begins with a set of requests from the contract and ends with
+one of those requests being handled.
 
 -}
 
--- | Generic event ID
-newtype EventId = EventId { unEventId :: UUID} -- stand-in for whatever we end up using
-  deriving  (Eq, Ord, Show, Generic)
-  deriving newtype (FromJSON, ToJSON)
+-- | Unique ID for contract instance
+newtype ContractInstanceId = ContractInstanceId { unContractInstanceId :: UUID }
+    deriving  (Eq, Ord, Show, Generic)
+    deriving newtype (FromJSON, ToJSON)
+    deriving Pretty via (PrettyShow UUID)
 
-data RequestEvent e =
-  IssueRequest EventId e
-  | CancelRequest EventId
-  deriving  (Eq, Ord, Show, Generic)
-  deriving anyclass (FromJSON, ToJSON)
+-- | How many times the contract has been advanced (how many events have been
+--   fed to it)
+newtype ContractIteration = ContractIteration { unContractIteration :: Int }
+    deriving  (Eq, Ord, Show, Generic)
+    deriving newtype (FromJSON, ToJSON, Pretty, Enum)
 
-data ResponseEvent e =
-  ResponseEvent
-      { respRequestId :: EventId
-      , respPayload   :: e
-      }
-  deriving  (Eq, Ord, Show, Generic)
-  deriving anyclass (FromJSON, ToJSON)
+iterationZero :: ContractIteration
```

Comment:
You *could* give it a monoid and make this mempty. Not sure if that's clearer tbh.

Reply:
I did add a monoid but I also kept `iterationZero`. I think I like the more expressive name here.